tsdb/head.go

```diff
@@ -1382,30 +1431,59 @@ func (a *headAppender) log() error {
 			return errors.Wrap(err, "log samples")
 		}
 	}
+	if len(a.exemplars) > 0 {
+		rec = enc.Exemplars(exemplarsForEncoding(a.exemplars), buf)
+		buf = rec[:0]
+
+		if err := a.head.wal.Log(rec); err != nil {
+			return errors.Wrap(err, "log exemplars")
+		}
+	}
 	return nil
 }
 
+func exemplarsForEncoding(es []exemplarWithSeriesRef) []record.RefExemplar {
+	ret := make([]record.RefExemplar, 0, len(es))
+	for _, e := range es {
+		ret = append(ret, record.RefExemplar{
+			Ref:    e.ref,
+			T:      e.exemplar.Ts,
+			V:      e.exemplar.Value,
+			Labels: e.exemplar.Labels,
+		})
+	}
+	return ret
+}
+
 func (a *headAppender) Commit() (err error) {
 	if a.closed {
 		return ErrAppenderClosed
 	}
 	defer func() { a.closed = true }()
-	if err := a.log(); err != nil {
-		_ = a.Rollback() // Most likely the same error will happen again.
-		return errors.Wrap(err, "write to WAL")
-	}
 
-	// No errors logging to WAL, so pass the exemplars along to the in memory storage.
+	// We need to know if an exemplar is going to be accepted into the exemplar storage before we write it to the WAL.
+	// This duplicating of (most) exemplars might not be as memory efficient as we'd like, but it's more time efficient
+	// than deleting from a slice in place.
+	commitExemplars := make([]exemplarWithSeriesRef, 0, len(a.exemplars))
 	for _, e := range a.exemplars {
 		s := a.head.series.getByID(e.ref)
-		// We don't instrument exemplar appends here, all is instrumented by storage.
-		if err := a.exemplarAppender.AddExemplar(s.lset, e.exemplar); err != nil {
-			if err == storage.ErrOutOfOrderExemplar {
-				continue
-			}
-			level.Debug(a.head.logger).Log("msg", "Unknown error while adding exemplar", "err", err)
+		err := a.exemplarAppender.AddExemplar(s.lset, e.exemplar)
+		if err == nil {
+			commitExemplars = append(commitExemplars, e)
+			continue
+		}
+
+		if err == storage.ErrOutOfOrderExemplar {
 			continue
 		}
+		level.Debug(a.head.logger).Log("msg", "Unknown error while adding exemplar", "err", err)
```

Comment:
Shouldn't it be a strong error?

Reply:
I'll set it as `Warn`, since we don't error out of `Commit` on an error here.